icu4c/source/common/ucnv.cpp

```diff
@@ -1803,10 +1811,18 @@ ucnv_toUChars(UConverter *cnv,
     }
     if(srcLength>0) {
         srcLimit=src+srcLength;
+
+        // Trim destCapacity such that computing the destLimit cannot
+        // overflow and wrap the pointer around. Which would be undefined behavior,
+        // meaning that compilers can optimize away attempts to check for it.
+        int32_t pinnedDestCapacity = (UChar *)U_MAX_PTR(dest) - (UChar *)dest;
```

Comment:
(UChar *)U_MAX_PTR(dest)     This cast is a little weird because the alignment of U_MAX_PTR may not be appropriate for (UChar).